hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/MergeIntoHoodieTableCommand.scala

```diff
@@ -160,7 +167,7 @@ case class MergeIntoHoodieTableCommand(mergeInto: MergeIntoTable) extends Hoodie
 
       // column order changed after left anti join , we should keep column order of source dataframe
       val cols = removeMetaFields(sourceDF).columns
-      executeInsertOnly(insertSourceDF.select(cols.head, cols.tail:_*), parameters)
+      executeInsertOnly(insertSourceDF.select(cols.head, cols.tail:_*), writeParam)
```

Comment:
this is basically saying if user sets precombine field, we always upsert, even if there's no match. i don't think this is the right semantics. if you want to de-duplicate the incoming records, then use `hoodie.combine.before.insert`.

Reply:
yes, use `hoodie.combine.before.insert` will de-duplicate, but this is not friendly to users.
When create a table with precombine field and use merge into sql to upsert data, it may be prod duplicate records if user wirte diff merge sql. if user need solve it, we need set `hoodie.combine.before.insert` in one case which only has  no match branch. User will have doubt, a table with precombineKey in merge sql, sometime writing effect is `upsert` and sometime `insert`.

Reply:
@KnightChess I think this pr wants to guarantee write consistency no matter whether the `when matched then` clause is present or not.
So maybe better to split this UT to two, one has configured `preCombineField`, and another not. And both of UT contains two cases that have `when matched then` or not.

Reply:
@KnightChess i think in the case of merge into, if we can set `hoodie.combine.before.insert` to true when precombine field is set, and keep the operation type still `insert` to align with "when not matched then insert *" where `insert` was used. Changing operation type leads to implementation inconsistency - even the method is called `executeInsertOnly()`

Reply:
@xushiyan I think `executeInsertOnly` and `executeUpsert` is different from hudi op `insert` and `upsert`, just a condition branch for `merge into` sql. And for the SQL Semantic, I think `merge into` shoudl only be used to `upsert` op, and event shoudle not follow the hudi `precombineKey`, because  `merget into` sql has a lot of flexibility to update the record which we want.

Reply:
@YannByron sorry, will add these days

Reply:
@KnightChess ok agree with the merge into semantics of doing upsert. To improve code readability, can we actually merge `executeUpsert` and `executeInsertOnly`? we're actually separate the condition flows within `executeUpsert`